collectd/collectd.go

```diff
@@ -0,0 +1,216 @@
+
+	// ErrServerClosed return when closing an already closed graphite server.
+	ErrServerClosed = errors.New("server already closed")
+
+	// ErrResolveUDPAddr returned when we are unable to resolve a udp address
+	ErrResolveUDPAddr = errors.New("unable to resolve UDP address")
+
+	// ErrListenUDP returned when we are unable to resolve a udp address
+	ErrListenUDP = errors.New("unable to listen on UDP")
+)
+
+// SeriesWriter defines the interface for the destination of the data.
+type SeriesWriter interface {
+	WriteSeries(database, retentionPolicy, name string, tags map[string]string, timestamp time.Time, values map[string]interface{}) error
+}
+
+type Server struct {
+	mu sync.Mutex
+	wg sync.WaitGroup
+
+	conn *net.UDPConn
+
+	writer      SeriesWriter
+	Database    string
+	typesdb     gollectd.Types
+	typesdbpath string
+}
+
+func NewServer(w SeriesWriter, typesDBPath string) *Server {
+	s := Server{
+		writer:      w,
+		typesdbpath: typesDBPath,
+		typesdb:     make(gollectd.Types),
+	}
+
+	return &s
+}
+
+func ListenAndServe(s *Server, iface string) error {
+	if iface == "" { // Make sure we have an address
+		return ErrBindAddressRequired
+	} else if s.Database == "" { // Make sure they have a database
+		return ErrDatabaseNotSpecified
+	}
+
+	addr, err := net.ResolveUDPAddr("udp", iface)
+	if err != nil {
+		// Should be something more like this, but testing becomes brittle
+		//return fmt.Errorf("unable to resolve UDP address: %v", err)
+		return ErrResolveUDPAddr
+	}
+
+	s.typesdb, err = gollectd.TypesDBFile(s.typesdbpath)
+	if err != nil {
+		// Should be something more like this, but testing becomes brittle
+		// return fmt.Errorf("unable to parse typesDBFile: %v", err)
+		return ErrCouldNotParseTypesDBFile
+	}
+
+	conn, err := net.ListenUDP("udp", addr)
+	if err != nil {
+		// Should be something more like this, but testing becomes brittle
+		//return fmt.Errorf("unable to listen on UDP: %v", err)
+		return ErrListenUDP
+	}
+	s.conn = conn
+
+	s.wg.Add(1)
+	go s.serve(conn)
+
+	return nil
+}
+
+func (s *Server) serve(conn *net.UDPConn) {
+	defer s.wg.Done()
+
+	// From https://collectd.org/wiki/index.php/Binary_protocol
+	//   1024 bytes (payload only, not including UDP / IP headers)
+	//   In versions 4.0 through 4.7, the receive buffer has a fixed size
+	//   of 1024 bytes. When longer packets are received, the trailing data
+	//   is simply ignored. Since version 4.8, the buffer size can be
+	//   configured. Version 5.0 will increase the default buffer size to
+	//   1452 bytes (the maximum payload size when using UDP/IPv6 over
+	//   Ethernet).
+	buffer := make([]byte, 1452)
+
+	for {
+		n, _, err := conn.ReadFromUDP(buffer)
+		if err != nil {
+			if s.conn == nil {
+				// This error occured becase we closed the connection and this is expected behavior
+				return
+			}
+			log.Printf("Collectd ReadFromUDP error: %s", err)
+		}
+
+		// Read in data in a separate goroutine.
+		s.wg.Add(1)
+		go s.handleMessage(buffer[:n])
```

Comment:
We are in a for loop, and if we don't launch a go routine, we have to wait until we handle the message and write it to the database before we can continue processing udp messages.   Line 114 can't read more until the loop comes back around and handle message would block that right?